src/SonarScanner.MSBuild.Tasks/IsTestByReference.cs

```diff
@@ -74,7 +74,7 @@ public override bool Execute()
             }
             else
             {
-                TestReference = References.Select(ParseName).FirstOrDefault(x => x != null && TestAssemblyNames.Contains(x.ToUpperInvariant()));
+                TestReference = References.Select(ParseName).FirstOrDefault(x => x != null && TestAssemblyNames.Contains(x));
```

Comment:
_Note:_ S4NET does this once per project -> we don't have a performance problem here.

At this step (MsBuild task), the references are not resolved yet. String names are taken from the project file as-is. So using case-sensitive check here silently drops support of a feature that might look like a regression to the users.

Maintaining uppercase here and normalcase on the other places would be painful. I suggest to keep the normalcase here plus something like `.Select(ToUpperCase).ToHashSet()`.

Reply:
The recommended approach for comparing string is to use `aaa.Equals(bbb, StringComparison.OrdinalIgnoreCase)` with the appropriate `StringComparison` for the situation.
Or since you are comparing against the contents of a generic `HashSet` you can specify the `StringComparer` type to use in the constructor.

Forcing to upper/lower then comparing is not a good approach. It's inefficient (creates unnecessary copies) and does not work correctly for all natural languages.

Reply:
We can replace the `HashSet` with `SortedSet(StringComparer.OrdinalIgnoreCase)` as the list is short, but it's not a performance hotspot in this project (contrary to sonar-dotnet repo).

The issue with natural languages is not a concern in this specific case, as we're looking only for a short hardcoded list of strings.

Reply:
> The issue with natural languages is not a concern in this specific case

Perhaps true from a functional point of view, but it creates an additional burden for anyone maintaining the code in the future - they have to understand the specific context to know whether or not it matters. Using the recommended approach means they don't even have to ask the question.

I can't see any reason not to use the recommended approach; it's faster, shows the intent better, and avoids future maintainers having to double-check whether the non-recommended approach works in context.

Reply:
I think that either way we do it (case insensitive or not), we need to have the same approach everywhere (S4NET, sonar-dotnet, security C# FE) to avoid inconsistent behavior.

IMO we should first do a benchmark between various approaches and take a decision based on that.

Reply:
That's exactly the purpose of this thread. Currently, it introduces inconsistency. `sonar-dotnet` and `sonar-security` have normalized reference names, while S4NET is taking value from user input.

Reply:
> Note: S4NET does this once per project -> we don't have a performance problem here.

The point (as the PR name suggests) is to have the same check for project type detection as we have in `sonar-dotnet`. We don't want a different functionality here.

Reply:
If we take the same approach here (where we take un-normalized user input) as in the analyzers (where MSBuild normalizes the references), we'll need to document this in the S4NET wiki + docs (and maybe in the perf troubleshooting guide).

It may surprise users to see that MSBuild is treating references case insensitive and we treat them case sensitive.

Reply:
and I saw now the implementation which is case insensitive, all good